pkg/envoy/xds_server.go

```diff
@@ -1259,19 +1272,27 @@ func getHTTPRule(secretManager certificatemanager.SecretManager, h *api.PortRule
 						},
 					})
 				} else {
-					// Only header presence needed
-					headers = append(headers, &envoy_config_route.HeaderMatcher{
-						Name:                 hdr.Name,
-						HeaderMatchSpecifier: &envoy_config_route.HeaderMatcher_PresentMatch{PresentMatch: true},
+					// Value is empty, but a secret is set, so we need to set an SDS value for it instead
+					log.Debugf("HeaderMatches: Adding %s because SDS value is required", hdr.Name)
+					headerMatches = append(headerMatches, &cilium.HeaderMatch{
+						MismatchAction: mismatch_action,
+						Name:           hdr.Name,
+						ValueSdsSecret: namespacedNametoSyncedSDSSecretName(types.NamespacedName{
+							Namespace: hdr.Secret.Namespace,
+							Name:      hdr.Secret.Name,
+						}, policySecretsNamespace),
 					})
+					// If we are supposed to fail on mismatch, we need to check on the presence of the
+					// header here.
+					if mismatch_action == cilium.HeaderMatch_FAIL_ON_MISMATCH {
+						// Check for header presence in headers
+						headers = append(headers, &envoy_config_route.HeaderMatcher{
+							Name:                 hdr.Name,
+							HeaderMatchSpecifier: &envoy_config_route.HeaderMatcher_PresentMatch{PresentMatch: true},
+						})
+					}
+
```

Comment:
IMO it is wrong, or at least very confusing to transform a HeaderMatch into two matches in Envoy like above when mismatch action is "fail-on-mismatch".

The old code treated fail-on-mismatch mismatch action separately as it is the default semantic of a regular Envoy HeaderMatcher. If we want to preserve full backwards compatibility when `hdr.Secret == nil` we should retain the old code as is. Alternatively, we could always translate also fail-on-mismatch mismatch action to `cilium.HeaderMatch` as it should behave the same.

Reply:
That's _much_ better, thanks! Applied, passes all the tests, including the connectivity ones.